lesson5 Databases

maximastashkin opened the pull request.

maximastashkin committed:
Some structure

maximastashkin committed:
Fixes by remarks. Added headIndex and tailIndex. New implementation of adding and removing elements

maximastashkin committed:
Fixed bug with an extra increase of the list size

maximastashkin committed:
lesson4 init. Collections p2

maximastashkin committed:
.gitignore

maximastashkin committed:
Add packages

maximastashkin committed:
Change getQuantityBy for .count()

maximastashkin committed:
Fixes after review

maximastashkin committed:
lesson5 init commit

maximastashkin committed:
Close connection fix

Add close() in Client class and use it in App for demo
Add /out/ folder to .gitignore

aleksin-rzn commented on src/main/kotlin/App.kt:
```diff
@@ -0,0 +1,80 @@
+    val client = Client("src/main/resources/test.db")
+    val sb = StringBuilder()
+
+    try {
```
Обработку исключений стоит вынести в методы классов реализующих их. В сообщении об ошибке стоит указать место возникновения исключения и текст ошибки

maximastashkin replied:
Можно немного поподробнее про обработку исключений. То есть если у меня методы в классе Client выбрасывают исключение или пробрасывают его, то обработку следует выполнять в классе DataGetter? И что тогда в catch делать? Организовывать вывод в поток не очень затея) Каким образом указывать место возникновения?

aleksin-rzn commented on src/main/kotlin/App.kt:
```diff
@@ -0,0 +1,80 @@
+        Initializer.createTables(client)
+
+    } catch (e: SQLException) {
+        sb.append(e.message).append("\n")
```
Чтобы не писать постоянно sb... нужно использовать scope-функцию with(sb){ ... }

aleksin-rzn commented on src/main/kotlin/DataGetter.kt:
```diff
@@ -0,0 +1,139 @@
+    DESCENDING("DESC");
+}
+
+class DataGetter {
```
Название ни о чем не говорит. Скорее всего это сервис такси или таксопарк.

aleksin-rzn commented on src/main/kotlin/DataGetter.kt:
```diff
@@ -0,0 +1,139 @@
+             WHERE id = ? """.trimIndent()
+
+            val result = client.executeQuery(sql, id.toString())
+            return Driver(result[0]["id"] as Int, result[0]["name"] as String, result[0]["drivingLicence"] as String)
```
result.first()["id"]

aleksin-rzn commented on src/main/kotlin/DataGetter.kt:
```diff
@@ -0,0 +1,139 @@
+
+            val result = client.executeQuery(sql, id.toString())
+            return DriverWithAllCars(
+                result[0]["id"] as Int,
```
result.first()

maximastashkin committed:
Fixes after review

1. Added with for sb: StringBuilder
2. Gave classes more descriptive names, added packages
3. Changed reference [0] to .first()
4. Reworked catching exceptions:
   -For exceptions that can be caught higher than in main, added handling => safe methods, which, if caught, return not instance, but null or an empty list. Unsafe methods still throw exceptions higher in function calls.
   -I also added my own class-exception for exceptions, which are thrown by the method that sends queries for updating with an indication of a more precise reason for throwing the exception and with information about the place in the call stack.

TakkaTooo commented:
https://github.com/TakkaTooo/tinkoff-kotlin-lessons/pull/4/commits/667334eb70c21376f6182f487067528f10505b10

maximastashkin committed:
Fixes throws exception

Now exceptions throws on update db methods throws by enum instance getting by errorCode

maximastashkin committed:
Fixes after review v2

Now important business information is logged before throwing the exception above.

aleksin-rzn commented on src/main/kotlin/App.kt:
```diff
@@ -4,47 +4,53 @@ import java.lang.StringBuilder
+                CarsharingInitializer.createAllTables(client)
+                CarsharingInitializer.fillAllTables(client)
+
+            } catch (e: CarsharingServiceOperationFaultException) {
```
1. Пустой catch очень плохой стиль
2. Если мы упали на создании таблиц или инициализации, то наверное нет смысла в дальнейшем выполнении программы и надо сразу идти в final вместе с нашей ошибкой, пытаться удалить таблицы, закрыть соединение и выкинуть в конце концов эту самую ошибку.
3. Попробуй runCatching вместо try/catch/finaly - тогда сможешь в scope-функциях let/also/apply работать как с успешным результатом из блока runCatching, так и с ошибкой.

aleksin-rzn commented on src/main/kotlin/App.kt:
```diff
@@ -4,47 +4,53 @@ import java.lang.StringBuilder
+            DataGetter.getDriversWithTripsAndCarInfo(client).forEach {
+                append(it).append("\n")
+            }
+        } catch (e: SQLException) {
```
Пустой catch

aleksin-rzn commented on src/main/kotlin/App.kt:
```diff
@@ -4,47 +4,53 @@ import java.lang.StringBuilder
+            // Removing
+            try {
+                CarsharingDropper.dropAllTables(client)
+            } catch (e: CarsharingServiceOperationFaultException) {
```
Пустой catch

aleksin-rzn commented on src/main/kotlin/App.kt:
```diff
@@ -4,47 +4,53 @@ import java.lang.StringBuilder
+            } catch (e: CarsharingServiceOperationFaultException) {
+            }
+            // Demo
+            append("Driver by id: ${DataGetter.getDriverById(client, 3)}").append("\n")
```
Почему DataGetter до сих пор не называется CarsharingService?

maximastashkin committed:
Removed empty catch's

Now the exception handling in main has been changed to a functional style.

TakkaTooo commented:
https://github.com/TakkaTooo/tinkoff-kotlin-lessons/pull/4/commits/7132bff6f91ebdfa1570a3ea6d17d8e94000de27

aleksin-rzn commented on src/main/kotlin/App.kt:
```diff
@@ -10,47 +10,40 @@ fun main() {
-            // Removing
-            try {
+        }.let {
+            println(this)
```
Вот так я печатал ошибку
```
}.let {
            if(it.isFailure){
                println(it.exceptionOrNull())
            }
```
На it.isFailure надо было печатать ошибку, на it.isSuccess можно было сообщить что все успешно

maximastashkin replied:
Не стал печать ошибку ещё раз, т.к. она логируется на более высоком уровне и потом пробрасывается выше ) т.е. в итоге сообщение об ошибке будет напечатано )
